### PR TITLE
replacer: allow to invoke from scripts

### PR DESCRIPTION
**Pull request by skirge**
```python
extReplacer = Control.getSingleton().getExtensionLoader().getExtension(ExtensionReplacer.NAME)
initiators = [ ]
replacer_rule = ReplacerParamRule(description=TOKEN_RULE_NAME,enabled=True,initiators=[],matchRegex=False,matchString="AuthenticationToken",matchType=MatchType.REQ_HEADER,replacement=token_value)
extReplacer.getParams().addRule(replacer_rule)
```

### CODE REVIEW

**Comment by psiinon** [comment]
Changes look good but you will also need to change https://github.com/zaproxy/zap-extensions/blob/beta/src/org/zaproxy/zap/extension/replacer/ZapAddOn.xml to increment the version and document the new change.

Thanks

**Comment by skirge** [comment]
Version incremented and description is added.

File: src/org/zaproxy/zap/extension/replacer/ExtensionReplacer.java [hunk 1]
```diff
@@ -96,7 +96,7 @@ private OptionsReplacerPanel getOptionsReplacerPanel() {
         return optionsReplacerPanel;
     }
 
-    protected ReplacerParam getParams() {
+    public ReplacerParam getParams() {
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/replacer/ExtensionReplacer.java]: Not sure exposing it is the best option, the rules are not immutable and not all of the changes done to them will be persisted, if we keep this change it's better to document that.

Comment (skirge) [reply src/org/zaproxy/zap/extension/replacer/ExtensionReplacer.java]: What do you suggest instead?

**Comment by psiinon** [comment]
This pull request **introduces 4 alerts** when merging cbe5bd435dd465fcce97b64662d6bf62e0e08f2b into e85eb86d6e5490a850b184d00caf809cb6614d79 - [view on LGTM.com](https://lgtm.com/projects/g/zaproxy/zap-extensions/rev/pr-ab0cd8a243bc0210369e5f9bbdf1b68e73e6689a)

**new alerts:**

* 3 for Unused variable, import, function or class
* 1 for Comparison between inconvertible types

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by thc202** [comment]
The pull request is now updated (after build/merge changes, zaproxy/zaproxy#5302).

**Comment by thc202** [comment]
Rebased to address the conflicts.

**Comment by psiinon** [comment]
Looks like an easy one to review - @kingthorin or @ricekot ?

**Comment by kingthorin** [comment]
I believe the concerns from Sept 2017 remain.

**Comment by psiinon** [comment]
Good point. So whats our proposed solution?
Can we suggest suitable words?
I've found the lack of visibility of lots of the ZAP methods a complete pain. I agree that exposing UI elements is not good but this one seems to be an ideal candidate for being public.
If we want suitable warnings then we can suggest some :)

**Comment by thc202** [comment]
The problem is not the lack visibility but lack of APIs. Just making the methods public will create maintenance problems, if the code is meant to be used by others it should provide an API to do so.

In this case it would be better that the extension allows to add/change/remove the rules instead of just exposing the `ReplacerParam`, that way we can still fix issues (bugs or enhancements) without worrying about breaking existing code. As for the wording, I'd suggest something like `Note: Not all changes done to the rules might be persisted nor all changes work as expected.`.

File: addOns/replacer/CHANGELOG.md [hunk 2]
```diff
@@ -4,6 +4,9 @@ All notable changes to this add-on will be documented in this file.
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
 
 ## Unreleased
+## Added
```
Comment (thc202) [inline addOns/replacer/CHANGELOG.md:7]: `### Added`, in any case this needs to be rebased.

**Comment by thc202** [comment]
Rebased and addressed the previous comment.

**Comment by kingthorin** [comment]
Still requires the overall change of functionality for add/update/remove rules vs just exposing the params. (Assuming I’m properly keeping track)

I don’t mind tackling that but it’ll be a few weeks at least as I’ve got some other items currently underway.

**Comment by thc202** [comment]
If we want to do that, yes, but I'm also ok leaving as is.

**Comment by kingthorin** [comment]
@psiinon tag you’re it 😉 which way would you like to see this go?

**Comment by psiinon** [comment]
Lets merge it - we can open new PRs for the extra functionality :)

**Review by psiinon — APPROVED** [review]

**Review by kingthorin — APPROVED** [review]

**Pull request merged**